### PR TITLE
[URP] Render Pass path store actions optimization

### PR DESCRIPTION
**Pull request by manuele-bonanno**
Fogbugz link:
https://fogbugz.unity3d.com/f/cases/1341452/

---
### Purpose of this PR
This PR introduces the same store actions optimization options introduced in [this PR](https://github.com/Unity-Technologies/Graphics/pull/4161) to the RenderPass path, to make sure we have performance parity between the RP and non-RP paths.

The main difference introduced is the concept of overriding Color/Depth store actions.
Since the RenderPass path merges render passes, it is important that the final store actions chosen for all the merged passes, matches the action requested **explicitly** by the last pass in the mergeable list.

All store actions are "Store" by default, the overridden flag helps to figure out which store actions are effectively requested by the user code and which ones were default store actions (which then are assumed safe to override in favour of the most optimal one)

---
### Testing status
Local, Yamato

---

### CODE REVIEW

**Comment by manuele-bonanno** [comment]
the few test failures are not related to this PR and the same as in master nightly run

**Review by jonuuukas — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/NativeRenderPass.cs [hunk 1]
```diff
@@ -135,6 +142,47 @@ internal void SetupNativeRenderPassFrameData(CameraData cameraData, bool isRende
             }
         }
 
+        internal void UpdateFinalStoreActions(int[] currentMergeablePasses, CameraData cameraData)
+        {
+            for (int i = 0; i < 8; ++i)
```
[the remark this answers is not in the document]
Comment (manuele-bonanno) [reply com.unity.render-pipelines.universal/Runtime/NativeRenderPass.cs:147]: I added this in a following PR, I think I could move that change here too and then merge later. WIll do

File: com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs [hunk 2]
```diff
@@ -163,6 +163,16 @@ public RenderBufferStoreAction depthStoreAction
             get => m_DepthStoreAction;
         }
 
+        public bool[] overriddenColorStoreActions
+        {
+            get => m_OverriddenColorStoreActions;
+        }
+
+        public bool overriddenDepthStoreAction
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs:171]: I'm failing to understand why this public API is needed.

Comment (manuele-bonanno) [reply com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs:171]: the fact that it is public is an error, will change that to internal. 
Regarding why the overridden flag is used I moved the conversation to slack

File: com.unity.render-pipelines.universal/Runtime/NativeRenderPass.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (manuele-bonanno) [reply com.unity.render-pipelines.universal/Runtime/NativeRenderPass.cs]: fixed in CS ad0a262

File: com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs (hunk 2 above)
Comment (manuele-bonanno) [reply com.unity.render-pipelines.universal/Runtime/Passes/ScriptableRenderPass.cs]: changed to internal in CS 5254f56

**Review by phi-lira — APPROVED** [review]

**Review by ernestasKupciunas — APPROVED** [review]
Looks fine. I was using [Winter](https://github.com/ernestasKupciunas/Winter) and [PolyFarm](https://github.com/Unity-Technologies/PolygonFarmBenchmark) projects to test the visual correctness and Polyfarm to test the performance. 

Performance doc:
On the performance side, I was comparing RP version build vs non-RP build with different store action configurations..
https://docs.google.com/spreadsheets/d/1zggujz-shSxvaw7rSfLT1hYVz0ZGIGMDHMSyvGX_HhM/edit?usp=sharing

Devices under testing:
iPad Air4, SoC: A14, iOS: 14.0
iPhone 12Pro, SoC: A14, iOS: 14.1
VLNQA00268, Samsung Galaxy S10+, 10.0.0, Exynos 9 9820, Mali-G76
OnePlus Nord, 10, Snapdragon 765G SM7250-AB, Adreno 620

Unity used for testing:
Version: 2021.2.0a20.2866
Revision: trunk 0bd21e314fca
Built: Thu, 03 Jun 2021 18:52:14 GMT

**Review by simon-engelbrecht-soerensen — APPROVED** [review]
Tested on PC with no difference in performance or visuals.

**Comment by manuele-bonanno** [comment]
after merging master into the branch seems like this is introducing a RenderPass test failure on Windows and Linux. All other test failures are not related. Investigation ongoing

**Comment by manuele-bonanno** [comment]
the remaining test errors are unrelated to this PR (out of memory and provisioning file issues on test devices). The 111_CameraStackMSAA_RenderPass test failure was fixed

**Pull request merged**